contracts/libs/MarginAccount.sol

```diff
@@ -1,12 +1,220 @@
+     */
+    struct Account {
+        //@dev the owner whose vaults we are looking for
+        address owner;
+        //@dev the number of vaults a user has, starting at index 0 and incrementing. vaultIds sorted chronologically.
+        uint256 vaultIds;
+    }
+
+    /**
+     * @dev Vault is a struct of 6 arrays that describe a position a user has. A user can have multiple vaults.
+     */
+    struct Vault {
+        /// @dev the addresses of oTokens a user has shorted (i.e. written) against this vault, including those in spreads/combos
+        address[] shortOtokens;
+        /**
+         * @dev the addresses of oTokens a user has gone long (i.e. bought) in this vault, including those in spreads/combos.
+         * Note that a user can be long oTokens without opening a vault (e.g. by buying on a DEX).
+         * Generally, long oTokens will be 'deposited' in vaults to act as collateral in order to write oTokens against (i.e. in spreads/combos).
+         */
+        address[] longOtokens;
+        /// @dev the addresses of other tokens a user has deposited as collateral in this vault
+        address[] collateralAssets;
+        /// @dev describes the quantity of options shorted for each oToken address in shortOtokens
+        uint256[] shortAmounts;
+        /// @dev describes the quantity of options longed in the vault for each oToken address in longOtokens
+        uint256[] longAmounts;
+        /// @dev describes the quantity of tokens deposited as collateral in the vault for each token address in collateralAssets
+        uint256[] collateralAmounts;
+    }
+
+    /**
+     * @dev Increment vault numbers in account
+     * @param _account The account relating to the user for whom the protocol is opening the new vault.
+     */
+    function _openNewVault(Account storage _account) internal {
+        _account.vaultIds = _account.vaultIds.add(1);
+    }
+
+    /**
+     * @dev increase the short oToken balance in a vault when a new oToken is minted.
+     * @param _vault The vault the protocol is adding the short position to
+     * @param _shortOtoken The address of the _shortOtoken the protocol is shorting from the user's vault
+     * @param _amount The additional number of _shortOtoken the protocol is shorting from the user's vault
+     * @param _index The index of _shortOtoken in the user's vault.shortOtokens array
+     */
+    function _addShort(
+        Vault storage _vault,
+        address _shortOtoken,
+        uint256 _amount,
+        uint256 _index
+    ) internal {
+        // Valid indexes in any array are between 0 and array.length - 1.
+        // If adding amount into an existant short otoken, make sure _index in the range of 0->length-1
+        if ((_index >= _vault.shortOtokens.length) && ((_index >= _vault.shortAmounts.length))) {
+            _vault.shortOtokens.push(_shortOtoken);
+            _vault.shortAmounts.push(_amount);
```

Comment:
so this case is when there is already a valid shortToken?

Reply:
hmm why is this case needed? Still confused

Reply:
@haythem96 maybe you can explain?

Reply:
yeah this is the case when there is no `shortOtokens` => `vault.shortOtokens.length ==0`, or when u are adding a new shortOtoken (e.g there exist one short, but u are adding a new one `vault.shortOtokens.length == 1, _index == 1`).

This case is needed, because when an array is empty `length == 0` the valid indices we can assign into is in the range of `0 and length-1`, that mean if we just do `vault.shortOtokens[index] = address` it will revert with `invalid opcode` message because we are accessing an invalid index., that's why we used `.push`.

Does that make sense ?

Reply:
similar to the case where an `array.length == 2` and we want to add a third element in the index `2`, the available indices we can access using `array[index] = something` are between 0 and `length-1 == 2-1= 1` so in that case will revert, that's why used `.push`

Reply:
In the else block, it basically mean that we are passing an already available index, so we can access without the need of creating a new one using `.push`.